test-framework/sudo-compliance-tests/src/sudoedit.rs

```diff
@@ -0,0 +1,432 @@
+};
+
+use crate::{Result, GROUPNAME, PANIC_EXIT_CODE, SUDOERS_ALL_ALL_NOPASSWD, USERNAME};
+
+#[cfg(not(target_os = "freebsd"))]
+const DEFAULT_EDITOR: &str = "/usr/bin/editor";
+#[cfg(target_os = "freebsd")]
+const DEFAULT_EDITOR: &str = "/usr/bin/vi";
+const LOGS_PATH: &str = "/tmp/logs.txt";
+const CHMOD_EXEC: &str = "555";
+const EDITOR_DUMMY: &str = "#!/bin/sh
+echo \"#\" >> \"$1\"";
+
+#[test]
+fn default_editor_is_usr_bin_editor() {
+    let expected = "default editor was called";
+    let env = Env(SUDOERS_ALL_ALL_NOPASSWD)
+        .user(USERNAME)
+        .file(
+            DEFAULT_EDITOR,
+            TextFile(format!(
+                "#!/bin/sh
+
+echo '{expected}' > {LOGS_PATH}"
+            ))
+            .chmod(CHMOD_EXEC),
+        )
+        .build();
+
+    Command::new("sudoedit")
+        .as_user(USERNAME)
+        .arg("/bin/foo.sh")
+        .output(&env)
+        .assert_success();
+
+    let actual = Command::new("cat").arg(LOGS_PATH).output(&env).stdout();
+
+    assert_eq!(expected, actual);
+}
+
+#[test]
+fn creates_sudoers_file_with_default_ownership_and_perms_if_it_doesnt_exist() {
+    let env = Env(SUDOERS_ALL_ALL_NOPASSWD)
+        .user(USERNAME)
+        .file(DEFAULT_EDITOR, TextFile(EDITOR_DUMMY).chmod(CHMOD_EXEC))
+        .build();
+
+    Command::new("rm")
+        .args(["-f", LOGS_PATH])
+        .output(&env)
+        .assert_success();
+
+    Command::new("sudoedit")
+        .arg("/foo.txt")
+        .as_user(USERNAME)
+        .output(&env)
+        .assert_success();
+
+    let ls_output = Command::new("ls")
+        .args(["-l", "/foo.txt"])
+        .output(&env)
+        .stdout();
+
+    assert_ls_output(&ls_output, "-rw-r--r--", "root", ROOT_GROUP);
+}
+
+#[test]
+fn passes_temporary_file_to_editor() {
+    let env = Env(SUDOERS_ALL_ALL_NOPASSWD)
+        .user(USERNAME)
+        .file(
+            DEFAULT_EDITOR,
+            TextFile(format!(
+                r#"#!/bin/sh
+echo "$@" > {LOGS_PATH}"#
+            ))
+            .chmod(CHMOD_EXEC),
+        )
+        .build();
+
+    Command::new("sudoedit")
+        .arg("/foo.txt")
+        .as_user(USERNAME)
+        .output(&env)
+        .assert_success();
+
+    let args = Command::new("cat").arg(LOGS_PATH).output(&env).stdout();
+
+    if sudo_test::is_original_sudo() {
+        assert_starts_with!(args, "/var/tmp");
+    } else {
+        assert_starts_with!(args, "/tmp");
+    }
+}
+
+#[test]
+fn temporary_file_owner_and_perms() {
+    let editor_script = format!(
+        r#"#!/bin/sh
+ls -l "$1" > {LOGS_PATH}"#
+    );
+
+    let env = Env(SUDOERS_ALL_ALL_NOPASSWD)
+        .group(GROUPNAME)
+        .user(USERNAME)
+        .file(DEFAULT_EDITOR, TextFile(editor_script).chmod(CHMOD_EXEC))
+        .build();
+
+    Command::new("sudoedit")
+        .arg(ETC_SUDOERS)
+        .as_user(USERNAME)
+        .output(&env)
+        .assert_success();
+
+    let ls_output = Command::new("cat").arg(LOGS_PATH).output(&env).stdout();
```

Comment:
Maybe we should have a method on `env` which reads a given file? Internally it could use `cat` and use `.assert_success()` to ensure that an error is given if the file doesn't exist. It would reduce all this duplication and in the future may be able to use `docker cp` to avoid the overhead of entering the namespaces of the container (if that overhead is actually noticable).